metric_learn/scml.py

```diff
@@ -0,0 +1,596 @@
+      n_basis = basis.shape[0]
+      if basis.shape[1] != n_features:
+        raise ValueError('The input dimensionality ({}) of the given '
+                         'linear transformation `init` must match the '
+                         'dimensionality of the given inputs `X` ({}).'
+                         .format(basis.shape[1], n_features))
+    elif self.basis not in authorized_basis:
+      raise ValueError(
+          "`basis` must be '{}' "
+          "or a numpy array of shape (n_basis, n_features)."
+          .format("', '".join(authorized_basis)))
+    if self.basis is authorized_basis[0]:
+      basis, n_basis = self._generate_bases_dist_diff(triplets, X)
+
+    return basis, n_basis
+
+  def _generate_bases_dist_diff(self, triplets, X):
+    """ Bases are generated from triplets as differences of positive or
+    negative pairs
+    TODO: complete function description
+    """
+
+    # TODO: Have a proportion of drawn pos and neg pairs?
+
+    # get all positive and negative pairs with lowest index first
+    # np.array (2*lenT,2)
+    T_pairs_sorted = np.sort(np.vstack((triplets[:, [0, 1]],
+                                        triplets[:, [0, 2]])),
+                             kind='stable')
+    # calculate all unique pairs and their indices
+    uniqPairs = np.unique(T_pairs_sorted, axis=0)
+
+    if self.n_basis is None:
+      # TODO: Get a good default n_basis directive
+      n_basis = uniqPairs.shape[0]
+      warnings.warn('The number of basis will be set to n_basis= %d' % n_basis)
+
+    elif isinstance(self.n_basis, int):
+      n_basis = self.n_basis
+    else:
+      raise ValueError("n_basis should be an integer, instead it is of type %s"
+                       % type(self.n_basis))
+
+    if n_basis > uniqPairs.shape[0]:
+      n_basis = uniqPairs.shape[0]
+      warnings.warn("The selected number of basis is greater than the number "
+                    "of points, only n_basis = %d will be generated" %
+                    n_basis)
+
+    uniqPairs = X[uniqPairs]
+
+    rng = check_random_state(self.random_state)
+
+    # Select n_basis
+    selected_pairs = uniqPairs[rng.choice(uniqPairs.shape[0],
+                               size=n_basis, replace=False), :, :]
+
+    basis = selected_pairs[:, 0]-selected_pairs[:, 1]
+
+    return basis, n_basis
+
+
+class SCML_global(_BaseSCML_global, _TripletsClassifierMixin):
+  """Sparse Compositional Metric Learning (SCML)
+
+  `SCML` builds a metric as the sparse positive combination of a set of locally
+  discriminative rank-one PSD basis. This allows an optimization scheme with
+  only `K` parameters, that can be yielded with an efficient stochastic
+  composite optimization over a set of triplets constraints. Each triplet is
+  constructed as a relative distance comparison with respect to the first
+  element so that the second element is closer than the last.
+  Read more in the :ref:`User Guide <scml>`.
+  Parameters
```

Comment:
Please use same line breaks convention as in docstring for other algorithms/functions

Reply:
Also indent conventions (i.e., 2 spaces) ;-)